src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const finalStates = [];
+  const stateClone = {};
+
+  Object.assign(stateClone, state);
+
+  for (const step in actions) {
+    switch (actions[step].type) {
+      case 'addProperties':
+        Object.assign(stateClone, actions[step].extraData);
+        break;
+      case 'removeProperties':
+        for (const key in actions[step].keysToRemove) {
+          const removeKey = actions[step].keysToRemove[key];
+
+          delete stateClone[removeKey];
+        }
+        break;
+
+      default:
+        for (const key in stateClone) {
+          delete stateClone[key];
+        }
+    }
+
+    finalStates.push(Object.assign({}, stateClone));
```

Comment:
The line `finalStates.push(Object.assign({}, stateClone));` should be placed outside the `switch` block but inside the loop, as per the prompt requirements. This ensures a copy of the state is pushed to the array after processing each action.